src/search/landmarks/landmark_count_heuristic.cc

```diff
@@ -263,20 +254,44 @@ bool LandmarkCountHeuristic::generate_helpful_actions(
 }
 
 bool LandmarkCountHeuristic::landmark_is_interesting(
-    const State &state, const LandmarkNodeSet &reached, LandmarkNode &lm_node) const {
-    /* A landmark is interesting if it hasn't been reached before and
-     its parents have all been reached, or if all landmarks have been
-     reached before, the LM is a goal, and it's not true at moment */
-
-    int num_reached = reached.size();
-    if (num_reached != lgraph->get_num_landmarks()) {
-        if (reached.find(&lm_node) != reached.end())
-            return false;
-        else
-            return !check_node_orders_disobeyed(lm_node, reached);
+    const State & /*state*/, LandmarkNode &lm_node) const {
+    /*
+      A landmark is interesting if it has to be achieved in the future and all
+      its parents have been reached.
+
+      TODO: Note that today, landmarks can only become needed-again after they
+       have been reached (in LAMA terms: accepted). This is only possible if all
+       their parents have been reached themselves. In that case, checking
+       whether all parents are reached in the switch case below seems
+       unnecessary, because *return true* sounds like the better option.
+       However, this ought to change once we incorporate the landmark
+       progression theory from our (unpublished) paper (see also issue1036).
+       According to that theory, landmarks are always accepted when first
+       reached, independent of the status of their parents. Due to reasonable
+       orderings, they can then also become needed-again even if not all their
+       parents are reached. Then, we should also consider them interesting only
+       after all their parents are reached, because the definition of reasonable
+       orderings tells us that reaching them earlier requires us to make them
+       false to reach their (reasonable) parents but also requires them to be
+       reached again after all their (reasonable) parents have been reached (or
+       at the same time).
+    */
+
+    landmark_status status =
+        lm_status_manager->get_landmark_status(lm_node.get_id());
+
+    switch (status) {
+    case lm_not_reached:
+    case lm_needed_again:
+        return all_of(lm_node.parents.begin(), lm_node.parents.end(),
+                      [&](const pair<LandmarkNode *, EdgeType> &parent) {
+                          int parent_id = parent.first->get_id();
+                          return lm_status_manager->is_reached(parent_id);
+                      });
+    case lm_reached:
+    default:
+        return false;
     }
```

Comment:
I think a `switch` that only really looks for one specific case (albeit a disjunctive one) is more cleanly expressed as an `if`. I'm not sure the `all_of` is easier to understand/read than a direct `for` loop, but if you prefer it, fine.